File: pset6/crack.py
import cs50
import string
import crypt
import sys

def Check(stringLength, genString, inputString):

    if(stringLength == len(genString)):
        # hash plaintext to be checked against input
        encStr = crypt.crypt("".join(genString), "50")

        # check for a match
        if(encStr == inputString):
            print("".join(genString))
            sys.exit(0)

    return 1;


def CrackPassword(input):

    chars = []

    # convert string of lower+uppercase alphabet to list
    for letter in string.ascii_letters:
        chars.append(letter)

    # generate 1-character plaintexts
    stringLength = 1
    genString = ["a"]

    for letter in chars:
        genString[0] = letter

        # Check the crypted word
        Check(stringLength, genString, input)

    # generate 2-character plaintexts
    stringLength = 2
    genString = ["a", "a"]

    for letter in chars:
        genString[0] = letter

        for letter2 in chars:
            genString[1] = letter2

            # Check the crypted word
            Check(stringLength, genString, input)

    # generate 3-character plaintexts
    stringLength = 3
    genString = ["a", "a", "a"]

    for letter in chars:
        genString[0] = letter

        for letter2 in chars:
            genString[1] = letter2

            for letter3 in chars:
                genString[2] = letter3

                # Check the crypted word
                Check(stringLength, genString, input)

    # generate 4-character plaintexts
    stringLength = 4
    genString = ["a", "a", "a", "a"]

    for letter in chars:
        genString[0] = letter

        for letter2 in chars:
            genString[1] = letter2

            for letter3 in chars:
                genString[2] = letter3

                for letter4 in chars:
                    genString[3] = letter4

                    # Check the crypted word
                    Check(stringLength, genString, input)

    # if no matches are found
    print("""The answer you're looking for either has more than 4 characters,
        doesnt consist ONLY of alphabets, or doesn't exist!""")
    sys.exit(1)

def main():

    if(len(sys.argv)) != 2:
        print("Error: Only accepting one command-line argument!")
        return 1;

    else:
        CrackPassword(sys.argv[1])

if __name__ == "__main__":
    main()

File: pset2/crack/crack.c
#define _XOPEN_SOURCE
#include <unistd.h>
#include <string.h>
#include <cs50.h>
#include <stdio.h>

void CrackPassword(string input);
int Check(int stringLength, string genString, string inputString);


int main(int argc, string argv[]) {

    if(argc != 2) {
        
        printf("Error: Only accepting one command-line argument!");
        return 1;
        
    } else {
        
        CrackPassword(argv[1]);
        
    }
}

void CrackPassword(string input) {
    
    int numOfChars = 51;
     
    char chars[] = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z',
               'A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'};
   
    char genString[6];
    
    //generate 1-character plaintexts
    int stringLength = 1;
    
        for (int i = 0; i < numOfChars; i++) {
        
            genString[0] = chars[i];
            
            //Check the crypted word
            Check(stringLength, genString, input);
            
            
        }
    
    //generate 2-character plaintexts
    stringLength = 2;
    
        for(int i = 0; i < numOfChars; i++) {
            
            genString[0] = chars[i];
            
                for(int o = 0; o < numOfChars; o++) {
                
                    genString[1] = chars[o];
                    
                    //Check the crypted word
                    Check(stringLength, genString, input);
                    
                }
           
        }
       
    //generate 3-character plaintexts
    stringLength = 3;
    
        for(int i = 0; i < numOfChars; i++) {
            
            genString[0] = chars[i];
            
                for(int o = 0; o < numOfChars; o++) {
                
                    genString[1] = chars[o];
                    
                        for(int p = 0; p < numOfChars;p++) {
                            
                            genString[2] = chars[p];
                            
                            //Check the crypted word
                            Check(stringLength, genString, input);
                            
                        }
                }
           
        }
    
    //generate 4-character plaintexts
    stringLength = 4;
    
        for(int i = 0; i < numOfChars; i++) {
            
            genString[0] = chars[i];
            
                for(int o = 0; o < numOfChars; o++) {
                
                    genString[1] = chars[o];
                    
                        for(int p = 0; p < numOfChars; p++) {
                            
                            genString[2] = chars[p];
                            
                                for(int l = 0; l < numOfChars; l++) {
                                    
                                    genString[3] = chars[l];
                                    
                                    //Check the crypted word
                                    Check(stringLength, genString, input);
                                    
                                }
                            
                        }
                        
                }
           
        }
        
    //If no matches are found
    printf("The answer you're looking for either has more than 4 characters, doesn't consist ONLY of alphabets, or doesn't exit!");
    exit(1);
}

int Check(int stringLength, string genString, string inputString) {
    
    if(stringLength == strlen(genString)) {
        
        //hash plaintext to be checked against input
        string encryptedStr = crypt(genString,"50");
        
        //check for a match
        if(strcmp(encryptedStr, inputString) == 0) {
            
            printf("%s\n", genString);
            
            exit(0);
        } 
        
    }
    return 1;
}